doc/designs/ech-api.md

```diff
@@ -0,0 +1,530 @@
+-------------------
+
+OpenSSL includes code for an
+[``sslecho``](https://github.com/sftcd/openssl/tree/ECH-draft-13c/demos/sslecho)
+demo.  We've added a minimal
+[``echecho``](https://github.com/sftcd/openssl/blob/ECH-draft-13c/demos/sslecho/echecho.c)
+that shows that adding one new server call
+(``SSL_CTX_ech_enable_server_buffer()``) and one new client call
+(``SSL_CTX_ech_set1_echconfig()``) is all that's needed to ECH-enable this
+demo.
+
+Handling Custom Extensions
+--------------------------
+
+OpenSSL supports custom extensions (via ``SSL_CTX_add_custom_ext()``) so that
+extension values are supplied and parsed by client and server applications via
+a callback.  The ECH specification of course doesn't deal with such
+implementation matters, but comprehensive ECH support for such custom
+extensions could quickly become complex. At present, in the absence of evidence
+of sensitive custom extension values, we handle all such extensions by using
+the ECH compression mechanism.  That means we require no API changes, only make
+one call to the application callbacks and get interoperability, but that such
+extension values remain visible to network observers. That could change if some
+custom value turns out to be sensitive such that we'd prefer to not include it
+in the outer CH.
+
+Server-side APIs
+----------------
+
+The main server-side APIs involve generating a key and the related
+ECHConfigList structure that ends up published in the DNS, periodically loading
+such keys into a server to prepare for ECH decryption and handling so-called
+ECH split-mode where a server only does ECH decryption but passes along the
+inner CH to another server that does the actual TLS handshake with the client.
+
+### Key and ECHConfigList Generation
+
+``ossl_edch_make_echconfig()`` is for use by command line or other key
+management tools, for example the ``openssl ech`` command documented
+[here](https://github.com/sftcd/openssl/blob/ECH-draft-13c/doc/man1/openssl-ech.pod.in).
+
+The ECHConfigList structure that will eventually be published in the DNS
+contains the ECH public value (an ECC public key) and other ECH related
+information, mainly the ``public_name`` that will be used as the SNI value in
+outer CH messages.
+
+```c
+int OSSL_ech_make_echconfig(unsigned char *echconfig, size_t *echconfiglen,
+                            unsigned char *priv, size_t *privlen,
+                            uint16_t ekversion, uint16_t max_name_length,
+                            const char *public_name, OSSL_HPKE_SUITE suite,
+                            const unsigned char *extvals, size_t extlen);
+```
+
+The ``echconfig`` and ``priv`` buffer outputs are allocated by the caller
+with the allocated size on input and the used-size on output. On output,
+the ``echconfig`` contains the base64 encoded ECHConfigList and the
+``priv`` value contains the PEM encoded PKCS#8 private value.
+
+The ``ekversion`` should be 0xfe0d for the current version.
```

Comment:
Mention the `OSSL_ECH_RFCXXXX_VERSION` macro here? Is there as yet no macro for the "current" ECH version?

Reply:
Ah, good point about a current version macro. Adding that now.